### PR TITLE
XTypes: merge from master

### PR DESCRIPTION
**Pull request by jrw972**

### CODE REVIEW

File: dds/DCPS/DataWriterImpl_T.h [hunk 1]
```diff
@@ -433,6 +450,43 @@ DataWriterImpl_T
     Message_Block_Ptr mb;
     ACE_Message_Block* tmp_mb;
 
+    if (marshal_skip_serialize_) {
+      size_t effective_size = 0, padding = 0;
+      if (marshaled_size_) {
+        effective_size = marshaled_size_;
+      } else {
+        // TODO:  Fix this.
```
Comment (jrw972) [inline dds/DCPS/DataWriterImpl_T.h:458]: @adamsj-oci This needs attention.

File: tests/transport/best_effort_reader/subscriber.cpp [hunk 2]
```diff
@@ -18,19 +19,37 @@ The test contains 3 writers and 3 readers, associated as follows:
...
+    using OpenDDS::RTPS::message_block_to_sequence;
+
+    ACE_INET_Addr remote_addr("127.0.0.1:12345");
+    OpenDDS::DCPS::LocatorSeq locators;
+    locators.length(1);
+    locators[0].kind = address_to_kind(remote_addr);
+    locators[0].port = remote_addr.get_port_number();
+    address_to_bytes(locators[0].address, remote_addr);
+
+    // TODO: Fix this.
```
Comment (jrw972) [inline tests/transport/best_effort_reader/subscriber.cpp:41]: Apply gen_find_size fix.

File: tests/transport/rtps/subscriber.cpp [hunk 3]
```diff
@@ -274,15 +277,31 @@ ACE_TMAIN(int argc, ACE_TCHAR* argv[])
...
+    using OpenDDS::RTPS::message_block_to_sequence;
+
+    ACE_INET_Addr remote_addr("127.0.0.1:12345");
+    LocatorSeq locators;
+    locators.length(1);
+    locators[0].kind = address_to_kind(remote_addr);
+    locators[0].port = remote_addr.get_port_number();
+    address_to_bytes(locators[0].address, remote_addr);
+
+    // TODO: Fix this.
```
Comment (jrw972) [inline tests/transport/rtps/subscriber.cpp:291]: Apply gen_find_size fix.

File: tests/transport/rtps_directed_write/subscriber.cpp [hunk 4]
```diff
@@ -150,13 +153,31 @@ void SimpleDataReader::data_received(const ReceivedDataSample& sample)
...
+    using OpenDDS::RTPS::message_block_to_sequence;
+
+    ACE_INET_Addr remote_addr("127.0.0.1:12345");
+    LocatorSeq locators;
+    locators.length(1);
+    locators[0].kind = address_to_kind(remote_addr);
+    locators[0].port = remote_addr.get_port_number();
+    address_to_bytes(locators[0].address, remote_addr);
+
+    // TODO: Fix this.
```
Comment (jrw972) [inline tests/transport/rtps_directed_write/subscriber.cpp:168]: Apply gen_find_size fix.

File: dds/DCPS/DataReaderImpl_T.h [hunk 5]
```diff
@@ -986,6 +987,16 @@ namespace OpenDDS {
     DataReaderImpl::qos_change(qos);
   }
 
+  void set_marshal_skip_serialize(bool value)
```
Comment (jwillemsen) [inline dds/DCPS/DataReaderImpl_T.h:990]: Why set/get?

File: dds/idl/marshal_generator.cpp [hunk 6]
```diff
@@ -2177,11 +2154,20 @@ namespace {
   bool generate_marshal_traits(
```
Comment (jrw972) [inline dds/idl/marshal_generator.cpp:2154]: @adamsj-oci Please check that I updated this code in a way that works.

File: dds/DCPS/DataWriterImpl_T.h [hunk 7]
```diff
@@ -15,6 +15,10 @@ OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
 namespace OpenDDS {
 namespace DCPS {
 
+enum {
+  cdr_header_size = 4
```
Comment (jwillemsen) [inline dds/DCPS/DataWriterImpl_T.h:19]: Just const?

File: dds/DCPS/Service_Participant.cpp [hunk 8]
```diff
@@ -1255,7 +1255,17 @@ Service_Participant::get_discovery(const DDS::DomainId_t domain)
       cf.open();
       ACE_Configuration_Section_Key k;
       cf.open_section(cf.root_section(), RTPS_SECTION_NAME, 1 /*create*/, k);
-      this->load_discovery_configuration(cf, RTPS_SECTION_NAME);
+
+      int status = load_discovery_configuration(cf, RTPS_SECTION_NAME);
```
Comment (jwillemsen) [inline dds/DCPS/Service_Participant.cpp:1259]: const

File: dds/DCPS/DataReaderImpl_T.h (hunk 5 above)
Comment (mitza-oci) [reply dds/DCPS/DataReaderImpl_T.h:990]: this is not something we're going to change here in the merge

File: dds/DCPS/DataWriterImpl_T.h (hunk 7 above)
Comment (iguessthislldo) [reply dds/DCPS/DataWriterImpl_T.h:19]: This should be removed.

**Comment by mitza-oci** [comment]
This PR exists only to merge `master` changes to `xtypes`.  There is not going to be a peer review here.  The discussions are those needed to resolve merge issues.

File: dds/DCPS/DataWriterImpl_T.h (hunk 1 above)
Comment (adamsj-oci) [reply dds/DCPS/DataWriterImpl_T.h:458]: @iguessthislldo suggests using `encoding_mode_.buffer_size(instance_data)`

Comment (jrw972) [reply dds/DCPS/DataWriterImpl_T.h:458]: Are we saying
```  ACE_Message_Block* dds_marshal(const MessageType& instance_data,
                                 OpenDDS::DCPS::MarshalingType marshaling_type)
  {
    const bool encapsulated = cdr_encapsulation();
    const Encoding& encoding = encoding_mode_.encoding();
    Message_Block_Ptr mb;
    ACE_Message_Block* tmp_mb;

    if (marshal_skip_serialize_) {
      const size_t effective_size = encoding_mode_.buffer_size(instance_data);
      ACE_NEW_MALLOC_RETURN(tmp_mb,
        static_cast<ACE_Message_Block*>(
          mb_allocator_->malloc(sizeof(ACE_Message_Block))),
        ACE_Message_Block(
          effective_size,
          ACE_Message_Block::MB_DATA,
          0, // cont
          0, // data
          data_allocator_.get(), // allocator_strategy
          get_db_lock(), // data block locking_strategy
          ACE_DEFAULT_MESSAGE_BLOCK_PRIORITY,
          ACE_Time_Value::zero,
          ACE_Time_Value::max_time,
          db_allocator_.get(),
          mb_allocator_.get()),
          0);
      mb.reset(tmp_mb);
      if (!MarshalTraitsType::to_message_block(*mb, instance_data)) {
        ACE_ERROR((LM_ERROR, ACE_TEXT("(%P|%t) ERROR: DataReaderImpl::dds_demarshal: ")
                   ACE_TEXT("attempting to skip serialize but bad from_message_block.\n")));
      }
      return mb.release();
    }
```

Comment (iguessthislldo) [reply dds/DCPS/DataWriterImpl_T.h:458]: Yes I think. You can put `encoding_mode_.buffer_size(instance_data)` into the constructor like the other ones.

Comment (adamsj-oci) [reply dds/DCPS/DataWriterImpl_T.h:458]: That's what I was expecting to see based on Fred's suggestion. There is a test `SkipSerialize` checking this so if that passes, then I'm good.

**Pull request merged**